### DISCUSSION TITLE
Custom binaries

### DISCUSSION THREAD

**Discussion by rothgar** [blinksh/blink#2014] [Ideas]
I have been using iSH for some of my local terminal needs but they have an odd i686 emulated environment which makes binary compatibility difficult. Looking at `uname -a` it appears Blink is ARM64 with a standard Darwin kernel.

Is there any chance you could support custom binaries in blink or maybe `brew`?

I can download a binary via curl but there's no `chmod` on the system to set it executable. I also can't find where binaries are stored because there's no `echo` to look at my $PATH

**Comment by carloscabanero** [comment]
Thanks a lot for the idea! As things are today, and this may change for WWDC, Blink is not able to execute arbitrary binaries, which is the reason why iSH uses an emulator. The local tooling we provide is just a convenience in order to let you do some tasks before jumping to a remote machine. You can think of it like a really simple BusyBox. The whole set of tools are running under a single process over multiple threads, and built as a library (something taken to the next level on http://github.com/holzschu/ios_system). That is why we also want to keep things simple and in a way that we can maintain easily.

Our goal with Blink is to make remote development feel like local.

That being said, I've been keeping an eye on Wasm and on how far we could take emulation in the future for "local" use-cases, maybe even running some isolated sandbox within the system for some simple tooling. But I haven't found anything that is there already. You may know more than I do on this though!

**Comment by rothgar** [comment]
I'd be interested in an emulation layer even if it isolates from the regular Blink interface and utilities. Something like `ssh localhost` could be very powerful and keep the stuff that can break separate from Blink itself.

I happily use Blink to connect to my remote systems, but often find myself without reliable network connectivity (planes, conferences, etc.) and having a toolbox of customizable local CLI applications would make blink stand out from what iSH provides. I've also been doing some conference talks directly from my iPad and Blink has been great for that so long as network works.

I was planning on running [UTM](https://getutm.app/) but it requires jailbraking for recent iPadOS versions which I'm not desparate enough to do yet. Maybe iOS alternate app stores will be available outside the EU at some point.
